profiles/sam-sa/terraform/sam/sa.tf

```diff
@@ -1,5 +1,97 @@
-# Create configuration bucket
-resource "google_service_account" "app_config" {
+# Application service account
+resource "google_service_account" "app" {
   account_id   = "${var.owner}-${var.service}"
   project      = "${var.google_project}"
 }
+
+resource "google_project_iam_member" "app" {
+  project = "${var.google_project}"
+  role    = "roles/compute.viewer"
+  member  = "serviceAccount:${google_service_account.app.email}"
+}
+
+resource "google_project_iam_member" "app_pubsub_binding" {
+  project = "${var.google_project}"
+  role    = "roles/pubsub.editor"
+  member  = "serviceAccount:${google_service_account.app.email}"
+}
+
+# Application Firestore service account
+resource "google_service_account" "firestore" {
+  account_id   = "${var.owner}-${var.service}-firestore"
+  project      = "${var.google_project}"
+}
+
+resource "google_project_iam_member" "firestore" {
+  project = "${var.google_project}"
+  role    = "roles/datastore.user"
+  member  = "serviceAccount:${google_service_account.firestore.email}"
+}
+
+# Admin SDK service accounts
+resource "google_service_account" "sam_sa_0" {
+  account_id   = "${var.owner}-${var.service}-0"
+  project      = "${var.google_project}"
+}
+resource "google_service_account" "sam_sa_1" {
+  account_id   = "${var.owner}-${var.service}-1"
+  project      = "${var.google_project}"
+}
+resource "google_service_account" "sam_sa_2" {
+  account_id   = "${var.owner}-${var.service}-2"
+  project      = "${var.google_project}"
+}
+
+# Grant service account access to container registry
+resource "google_storage_bucket_iam_member" "app" {
+  bucket = "${var.gcr_bucket_name}"
+  role   = "roles/storage.objectViewer"
+  member = "serviceAccount:${google_service_account.app.email}"
+}
+
+resource "google_service_account_key" "app_account_key" {
+  service_account_id = "${google_service_account.app.name}"
+}
+
+resource "google_service_account_key" "app_firestore_account_key" {
+  service_account_id = "${google_service_account.firestore.name}"
+}
+
+resource "google_service_account_key" "sam_sa_0_key" {
+  service_account_id = "${google_service_account.firestore.name}"
+}
+
+resource "google_service_account_key" "sam_sa_1_key" {
+  service_account_id = "${google_service_account.firestore.name}"
+}
+
+resource "google_service_account_key" "sam_sa_2_key" {
+  service_account_id = "${google_service_account.firestore.name}"
+}
```

Comment:
These could probably (?) use a `count`. I wonder why we need exactly 3 keys for the same SA.

Reply:
these should be separate SAs for separate directory accounts, not the firestore account. They need some kind of domain permissions...